### ISSUE TITLE
When trying to save an entry to the database, my app crashes.

### ISSUE THREAD

**Issue by rugaard** [vapor/vapor#522]
I'm currently using `Vapor 0.16` with `Fluent` and the latest `MySQL` provider.

When I'm trying to insert a new record into my database, my app crashes and returns the following stack trace in my Xcode:
```
POST /admin/backend-users/store
fatal error: Can't form Range with upperBound < lowerBound
Current stack trace:
0    libswiftCore.dylib                 0x000000010096c750 swift_reportError + 125
1    libswiftCore.dylib                 0x0000000100989120 _swift_stdlib_reportFatalError + 61
2    libswiftCore.dylib                 0x0000000100784b30 specialized specialized StaticString.withUTF8Buffer<A> ((UnsafeBufferPointer<UInt8>) -> A) -> A + 351
3    libswiftCore.dylib                 0x00000001008fb210 partial apply for (_fatalErrorMessage(StaticString, StaticString, StaticString, UInt, flags : UInt32) -> Never).(closure #2) + 109
4    libswiftCore.dylib                 0x0000000100784b30 specialized specialized StaticString.withUTF8Buffer<A> ((UnsafeBufferPointer<UInt8>) -> A) -> A + 351
5    libswiftCore.dylib                 0x00000001008b93e0 specialized _fatalErrorMessage(StaticString, StaticString, StaticString, UInt, flags : UInt32) -> Never + 96
6    Vapor                              0x00000001005dd170 Collection<A where ...>.split(separator : [UInt8], excludingFirst : Bool, excludingLast : Bool, maxSplits : Int?) -> [[UInt8]] + 2881
7    Vapor                              0x00000001005a0ac0 static Multipart.parse([UInt8], boundary : String) -> [String : Multipart] + 1079
8    Vapor                              0x00000001005ade70 Request.multipart.getter + 1720
9    Vapor                              0x000000010056e810 Request.(data.getter).(closure #4) + 767
10   Vapor                              0x000000010055bf50 thunk + 31
11   Vapor                              0x000000010055bfc0 partial apply for thunk + 81
12   Vapor                              0x000000010055d3a0 thunk + 39
13   Vapor                              0x000000010055d540 partial apply for thunk + 81
14   Vapor                              0x000000010055cec0 Content.(subscript.getter).(closure #1) + 82
15   Vapor                              0x000000010055cf70 partial apply for Content.(subscript.getter).(closure #1) + 81
16   Vapor                              0x000000010055cfd0 thunk + 129
17   Vapor                              0x000000010055d0a0 partial apply for thunk + 81
18   libswiftCore.dylib                 0x00000001007843d0 Optional.map<A> ((A) throws -> A1) throws -> A1? + 363
19   libswiftCore.dylib                 0x0000000100803950 LazyMapIterator.next() -> B? + 182
20   libswiftCore.dylib                 0x0000000100803aa0 protocol witness for IteratorProtocol.next() -> A.Element? in conformance <A, B where ...> LazyMapIterator<A, B> + 23
21   libswiftCore.dylib                 0x00000001007a9a30 LazyFilterIterator.next() -> A.Element? + 237
22   libswiftCore.dylib                 0x00000001007a9d00 protocol witness for IteratorProtocol.next() -> A.Element? in conformance <A where ...> LazyFilterIterator<A> + 23
23   libswiftCore.dylib                 0x0000000100760ae0 Collection.first.getter + 108
24   libswiftCore.dylib                 0x00000001007adbb0 protocol witness for Collection.count.getter in conformance <A where ...> LazyFilterCollection<A> + 156
25   libswiftCore.dylib                 0x00000001007adb80 protocol witness for Collection.first.getter in conformance <A where ...> LazyFilterBidirectionalCollection<A> + 28
26   libswiftCore.dylib                 0x0000000100805ab0 LazyMapBidirectionalCollection.first.getter + 194
27   Vapor                              0x000000010055caa0 Content.subscript.getter + 579
28   Vapor                              0x000000010055cda0 Content.subscript.getter + 172
29   VaporBackend                       0x0000000100726320 BackendUserController.store(Request) throws -> ResponseRepresentable + 130
30   VaporBackend                       0x0000000100728090 partial apply for BackendUserController.store(Request) throws -> ResponseRepresentable + 30
31   TypeSafeRouting                    0x00000001004bca10 RouteBuilder<A where ...>.(post(String, handler : (Request) throws -> ResponseRepresentable) -> ()).(closure #1) + 125
32   TypeSafeRouting                    0x00000001004bccb0 partial apply for RouteBuilder<A where ...>.(post(String, handler : (Request) throws -> ResponseRepresentable) -> ()).(closure #1) + 89
33   HTTPRouting                        0x00000001004abb30 RouteBuilder<A where ...>.(add(Method, [String], (Request) throws -> ResponseRepresentable) -> ()).(closure #1) + 90
34   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
35   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
36   Vapor                              0x000000010058a890 FileMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 171
37   Vapor                              0x000000010058b4f0 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance FileMiddleware + 83
38   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
39   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
40   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
41   Vapor                              0x00000001005c5e60 SessionMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 934
42   Vapor                              0x00000001005c6660 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance SessionMiddleware + 83
43   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
44   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
45   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
46   Vapor                              0x00000001005d45d0 ValidationMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 163
47   Vapor                              0x00000001005d4f40 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance ValidationMiddleware + 83
48   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
49   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
50   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
51   Vapor                              0x000000010057c540 DateMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 163
52   Vapor                              0x000000010057c910 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance DateMiddleware + 67
53   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
54   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
55   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
56   Vapor                              0x00000001005bcf60 TypeSafeErrorMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 118
57   Vapor                              0x00000001005bd270 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance TypeSafeErrorMiddleware + 67
58   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
59   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
60   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
61   Vapor                              0x000000010059ea40 AbortMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 174
62   Vapor                              0x000000010059fcf0 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance AbortMiddleware + 83
63   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
64   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
65   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
66   App                                0x0000000100005c00 SampleMiddleware.respond(to : Request, chainingTo : Responder) throws -> Response + 118
67   App                                0x0000000100005df0 protocol witness for Middleware.respond(to : Request, chainingTo : Responder) throws -> Response in conformance SampleMiddleware + 83
68   Vapor                              0x00000001005a05b0 Collection<A where ...>.(chain(to : Responder) -> Responder).(closure #1).(closure #1) + 227
69   HTTP                               0x00000001002f77c0 Request.Handler.respond(to : Request) throws -> Response + 82
70   HTTP                               0x00000001002f7870 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Request.Handler + 68
71   Vapor                              0x000000010057d290 Droplet.respond(to : Request) throws -> Response + 2029
72   Vapor                              0x000000010057ee60 protocol witness for Responder.respond(to : Request) throws -> Response in conformance Droplet + 59
73   HTTP                               0x00000001002fc580 Server.respond(stream : Stream, responder : Responder) throws -> () + 805
74   HTTP                               0x00000001002fc250 Server.(start(responder : Responder, errors : (ServerError) -> ()) throws -> ()).(closure #1) + 233
75   Core                               0x000000010023b150 thunk + 39
76   libswiftDispatch.dylib             0x0000000100c5ee10 partial apply for thunk + 36
77   libdispatch.dylib                  0x0000000100f55064 _dispatch_call_block_and_release + 12
78   libdispatch.dylib                  0x0000000100f47cbd _dispatch_client_callout + 8
79   libdispatch.dylib                  0x0000000100f4b8e1 _dispatch_root_queue_drain + 2934
80   libdispatch.dylib                  0x0000000100f4b83b _dispatch_worker_thread3 + 106
81   libsystem_pthread.dylib            0x0000000100fa9ecd _pthread_wqthread + 1129
82   libsystem_pthread.dylib            0x0000000100fa7f84 start_wqthread + 13
(lldb) 
```

Xcode also opens the file `Vapor/Utilities/Byte/Data.swift` and jumps down the the method `split()` and marks the following line as the cause of the crash:
```swift
parts.append(self[range.to..<nextRange.from].array)
```

---

I've followed the examples from the Vapor documentation and have created the following controller action, which when executed crashes the app:

```swift
/**
    Store user
     
    - parameter request: HTTPRequest
*/
public func store(_ request: Request) throws -> ResponseRepresentable { 
    // Create new user
    var user = BackendUser(
        name: request.data["name"].string ?? "",
        email: request.data["email"].string ?? "",
        password: request.data["password"].string ?? "",
        role: "developer"
    )
        
    // Save user to database
    try user.save()
        
    return try JSON([
        "data": user
    ])  
}
```

And the model I'm using to insert a database looks like this:
```swift
import Vapor
import Fluent
import Foundation

public final class BackendUserToken: Model {
    
    public static var entity = "backend_user_tokens"
    
    public var id: Node?
    public var user_id: Int
    public var token: String
    public var expires_at: Int?
    public var created_at: Int
    public var updated_at: Int
    
    public init(user_id: Int, token: String, expires_at: Int?, created_at: Int, updated_at: Int) {
        self.user_id = user_id
        self.token = token
        self.expires_at = expires_at
        self.created_at = created_at
        self.updated_at = updated_at
    }
    
    public convenience init(user_id: Int, token: String, expires_at: Int?) {
        self.init(
            user_id: user_id,
            token: token,
            expires_at: expires_at,
            created_at: Int(NSDate().timeIntervalSince1970),
            updated_at: Int(NSDate().timeIntervalSince1970)
        )
    }
    
    public convenience init(user_id: Int, token: String) {
        self.init(
            user_id: user_id,
            token: token,
            expires_at: nil,
            created_at: Int(NSDate().timeIntervalSince1970),
            updated_at: Int(NSDate().timeIntervalSince1970)
        )
    }
    
    public init(node: Node, in context: Context) throws {
        id = try node.extract("id")
        user_id = try node.extract("user_id")
        token = try node.extract("token")
        expires_at = try node.extract("expires_at")
        created_at = try node.extract("created_at")
        updated_at = try node.extract("updated_at")
    }
    
    public func makeNode() throws -> Node {
        return try Node(node: [
            "id": id,
            "user_id": user_id,
            "token": token,
            "expires_at": expires_at,
            "created_at": created_at,
            "updated_at": updated_at
            ])
    }
    
    /*---------------------------------------------
     | Preparations
     ---------------------------------------------*/
    
    /**
     Prepare database with entity table
     
     - parameter _database: Database
     */
    public static func prepare(_ database: Database) throws {
        try database.create(entity) { builder in
            builder.id()
            builder.int("user_id");
            builder.string("token", length: 60)
            builder.int("expires_at", optional: true)
            builder.int("created_at")
            builder.int("updated_at")
        }
    }
    
    /**
     Revert entity table from database
     
     - parameter _database: Database
     */
    public static func revert(_ database: Database) throws {
        try database.delete(entity)
    }
}
```

**Comment by tannernelson** [comment]
I'll get this fixed for 0.17.

**Comment by LoganWright** [comment]
I think this has been fixed for a while, let's reopen if it hasn't
